File: scripts/rust/src/review_impl_order.py
#!/usr/bin/env python3
"""
Review: Implementation order - standard traits should be at the bottom.

Malpattern:
1. Data structure (struct/enum)
2. Trait definition for that data structure
3. Standard trait impls (Eq, PartialEq, Debug, Display, etc.) <- WRONG POSITION
4. Inherent impl (impl Type { ... })
5. Custom trait impls

Correct order:
1. Data structure (struct/enum)
2. Trait definition for that data structure
3. Inherent impl (impl Type { ... })
4. Custom trait impls
5. Standard trait impls (Eq, PartialEq, Debug, Display, etc.) <- AT THE BOTTOM

Standard traits include: Eq, PartialEq, Ord, PartialOrd, Debug, Display, Clone, 
Copy, Hash, Default, From, Into, TryFrom, TryInto, Deref, DerefMut, Drop, etc.
"""
# Git commit: 584a672b6a34782766863c5f76a461d3297a741a
# Date: 2025-10-17 05:17:36 -0700


import re
import sys
from pathlib import Path

# Standard library traits that should come after custom impls
STANDARD_TRAITS = {
    'Eq', 'PartialEq', 'Ord', 'PartialOrd',
    'Debug', 'Display', 
    'Clone', 'Copy',
    'Hash', 
    'Default',
    'From', 'Into', 'TryFrom', 'TryInto',
    'AsRef', 'AsMut',
    'Deref', 'DerefMut',
    'Drop',
    'Iterator', 'IntoIterator',
    'Index', 'IndexMut',
    'Add', 'Sub', 'Mul', 'Div', 'Rem', 'Neg',  # Arithmetic ops
    'BitAnd', 'BitOr', 'BitXor', 'Shl', 'Shr',  # Bitwise ops
    'Not',
    'Send', 'Sync',
    'Fn', 'FnMut', 'FnOnce',
    'Error',
}


def extract_trait_name(impl_line):
    """
    Extract trait name from an impl line.
    Returns (trait_name, is_trait_impl)
    
    Examples:
    impl Eq for Foo -> ('Eq', True)
    impl<T> Display for Foo<T> -> ('Display', True)
    impl<T> std::fmt::Debug for Foo<T> -> ('Debug', True)
    impl<T: StT> MyTrait<T> for Foo<T> -> ('MyTrait', True)
    impl Foo -> (None, False)  # inherent impl
    impl<T> Foo<T> -> (None, False)  # inherent impl
    """
    # Remove comments
    impl_line = re.sub(r'//.*$', '', impl_line).strip()
    
    # Pattern: impl [generic_params] TraitName[<trait_generics>] for Type
    # or: impl [generic_params] Type  (inherent impl)
    
    # Check for trait impl: "impl ... Trait for Type"
    # Handle:
    # - qualified paths like std::fmt::Debug
    # - trait generics like MyTrait<T>
    match = re.search(r'impl(?:<[^>]+>)?\s+(?:[\w:]+::)?(\w+)(?:<[^>]+>)?\s+for\s+', impl_line)
    if match:
        trait_name = match.group(1)
        return (trait_name, True)
    
    # Inherent impl
    return (None, False)


def is_struct_or_enum_line(line):
    """Check if line defines a struct or enum."""
    stripped = line.strip()
    return bool(re.match(r'(pub\s+)?(?:struct|enum)\s+\w+', stripped))


def is_trait_definition(line):
    """Check if line defines a trait."""
    stripped = line.strip()
    return bool(re.match(r'(pub\s+)?trait\s+\w+', stripped))


def is_impl_line(line):
    """Check if line starts an impl block."""
    stripped = line.strip()
    return bool(re.match(r'impl(?:<[^>]+>)?\s+', stripped))


def check_impl_order(file_path):
    """
    Check for standard trait impls appearing before custom trait impls.
    Standard traits should be at the bottom of the file.
    
    Returns list of violations: (file_path, line_num, trait_name, context)
    """
    violations = []
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
    except Exception as e:
        print(f"Error reading {file_path}: {e}", file=sys.stderr)
        return violations
    
    # Track state for each data structure
    current_struct = None
    struct_line = None
    seen_custom_impl = False
    seen_standard_impl = False
    first_standard_impl = None
    
    for i, line in enumerate(lines, 1):
        stripped = line.strip()
        
        # Skip empty lines and comments
        if not stripped or stripped.startswith('//'):
            continue
        
        # New struct/enum resets state
        if is_struct_or_enum_line(line):
            # Reset for new struct
            match = re.search(r'(?:struct|enum)\s+(\w+)', stripped)
            if match:
                current_struct = match.group(1)
                struct_line = i
                seen_custom_impl = False
                seen_standard_impl = False
                first_standard_impl = None
            continue
        
        # Check impl blocks
        if is_impl_line(line):
            trait_name, is_trait_impl = extract_trait_name(line)
            
            if not is_trait_impl:
                # Inherent impl - ignore
                continue
            
            if not trait_name:
                continue
            
            # Determine if it's a standard or custom trait
            if trait_name in STANDARD_TRAITS:
                # Standard trait impl
                if not seen_standard_impl:
                    seen_standard_impl = True
                    first_standard_impl = (i, trait_name, stripped)
            else:
                # Custom trait impl
                if seen_standard_impl:
                    # Violation: standard impl came before this custom impl
                    violations.append({
                        'file': file_path,
                        'struct': current_struct,
                        'struct_line': struct_line,
                        'standard_impl_line': first_standard_impl[0],
                        'standard_trait': first_standard_impl[1],
                        'custom_impl_line': i,
                        'custom_trait': trait_name,
                    })
                
                seen_custom_impl = True
    
    return violations


def main():
    repo_root = Path(__file__).parent.parent.parent.parent
    
    search_dirs = [
        repo_root / "src",
    ]
    
    all_violations = []
    
    for search_dir in search_dirs:
        if not search_dir.exists():
            continue
        
        for rs_file in search_dir.rglob("*.rs"):
            violations = check_impl_order(rs_file)
            all_violations.extend(violations)
    
    if all_violations:
        print("✗ Implementation Order Violations:\n")
        print("Standard trait impls (Eq, Debug, Display, etc.) should be AT THE BOTTOM (after custom trait impls).\n")
        
        for v in all_violations:
            rel_path = v['file'].relative_to(repo_root)
            print(f"  {rel_path}:{v['struct_line']}")
            if v['struct']:
                print(f"    Struct: {v['struct']}")
            print(f"    Line {v['standard_impl_line']}: {v['standard_trait']} impl (standard trait)")
            print(f"    Line {v['custom_impl_line']}: {v['custom_trait']} impl (custom trait)")
            print(f"    → Standard trait impls should move to the bottom (after all custom impls)")
            print()
        
        print(f"Total violations: {len(all_violations)}")
        print("\nCorrect order:")
        print("  1. Data structure (struct/enum)")
        print("  2. Trait definition")
        print("  3. Inherent impl (impl Type { ... })")
        print("  4. Custom trait implementations")
        print("  5. Standard trait implementations (Eq, Debug, etc.) <- AT THE BOTTOM")
        return 1
    
    print("✓ All implementations are in correct order")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/bin/review_impl_order.rs
// Copyright (C) Brian G. Milnes 2025

//! Review: Implementation order - standard traits should be at the bottom
//! 
//! Replaces: scripts/rust/src/review_impl_order.py
//! Rule: Standard trait impls (Eq, Debug, Display, etc.) must come AFTER custom trait impls
//! Git commit: 584a672b6a34782766863c5f76a461d3297a741a
//! Binary: rusticate-review-impl-order
//!
//! Correct order:
//! 1. Data structure (struct/enum)
//! 2. Trait definition
//! 3. Inherent impl (impl Type { ... })
//! 4. Custom trait implementations
//! 5. Standard trait implementations <- AT THE BOTTOM

use anyhow::Result;
use rusticate::{StandardArgs, format_number, parse_source, find_nodes, find_rust_files};
use ra_ap_syntax::{SyntaxKind, ast::{self, AstNode}};
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;


macro_rules! log {
    ($($arg:tt)*) => {{
        use std::io::Write;
        let msg = format!($($arg)*);
        println!("{}", msg);
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("analyses/review_impl_order.log")
        {
            let _ = writeln!(file, "{}", msg);
        }
    }};
}
fn is_standard_trait(name: &str) -> bool {
    matches!(name,
        "Eq" | "PartialEq" | "Ord" | "PartialOrd" |
        "Debug" | "Display" |
        "Clone" | "Copy" |
        "Hash" |
        "Default" |
        "From" | "Into" | "TryFrom" | "TryInto" |
        "AsRef" | "AsMut" |
        "Deref" | "DerefMut" |
        "Drop" |
        "Iterator" | "IntoIterator" |
        "Index" | "IndexMut" |
        "Add" | "Sub" | "Mul" | "Div" | "Rem" | "Neg" |
        "BitAnd" | "BitOr" | "BitXor" | "Shl" | "Shr" |
        "Not" |
        "Send" | "Sync" |
        "Fn" | "FnMut" | "FnOnce" |
        "Error"
    )
}

#[derive(Debug)]
struct Violation {
    file: PathBuf,
    standard_line: usize,
    standard_trait: String,
    custom_line: usize,
    custom_trait: String,
}

fn check_file(file_path: &Path, source: &str) -> Result<Vec<Violation>> {
    let source_file = parse_source(source)?;
    let root = source_file.syntax();
    
    // Find all IMPL nodes
    let impl_nodes = find_nodes(root, SyntaxKind::IMPL);
    
    // Track impl order
    let mut impls = Vec::new();
    
    for impl_node in impl_nodes {
        if let Some(impl_ast) = ast::Impl::cast(impl_node.clone()) {
            // Check if this is a trait impl (has trait_() method)
            if let Some(trait_type) = impl_ast.trait_() {
                // Get trait name
                let trait_text = trait_type.syntax().text().to_string();
                let trait_name = trait_text
                    .split(|c: char| !c.is_alphanumeric() && c != '_')
                    .find(|s| !s.is_empty())
                    .unwrap_or(&trait_text)
                    .to_string();
                
                let line_num = rusticate::line_number(impl_ast.syntax(), source);
                let is_standard = is_standard_trait(&trait_name);
                
                impls.push((line_num, trait_name, is_standard));
            }
        }
    }
    
    // Check for violations: standard trait impl before custom trait impl
    let mut violations = Vec::new();
    let mut first_standard: Option<(usize, String)> = None;
    
    for (line_num, trait_name, is_standard) in impls {
        if is_standard {
            if first_standard.is_none() {
                first_standard = Some((line_num, trait_name));
            }
        } else {
            // Custom trait impl
            if let Some((std_line, ref std_trait)) = first_standard {
                // Violation: standard impl came before this custom impl
                violations.push(Violation {
                    file: file_path.to_path_buf(),
                    standard_line: std_line,
                    standard_trait: std_trait.clone(),
                    custom_line: line_num,
                    custom_trait: trait_name,
                });
            }
        }
    }
    
    Ok(violations)
}

fn main() -> Result<()> {
    let start = Instant::now();
    let args = StandardArgs::parse()?;
    let base_dir = args.base_dir();
    
    // Print compilation directory for Emacs compile-mode
    log!("Entering directory '{}'", base_dir.display());
    log!("");
    
    let search_dirs = args.get_search_dirs();
    
    // This rule only applies to src/ (not tests/ or benches/)
    let src_dirs: Vec<_> = search_dirs.iter()
        .filter(|p| p.is_dir() && (p.ends_with("src") || p.components().any(|c| c.as_os_str() == "src")))
        .cloned()
        .collect();
    
    if src_dirs.is_empty() {
        log!("✓ No src/ directories to check");
        let elapsed = start.elapsed().as_millis();
        log!("Completed in {}ms", elapsed);
        return Ok(());
    }
    
    let files = find_rust_files(&src_dirs);
    let mut all_violations = Vec::new();
    
    for file in &files {
        match fs::read_to_string(file) {
            Ok(source) => {
                match check_file(file, &source) {
                    Ok(violations) => all_violations.extend(violations),
                    Err(e) => {
                        eprintln!("Warning: Failed to parse {}: {}", file.display(), e);
                    }
                }
            }
            Err(e) => {
                eprintln!("Warning: Failed to read {}: {}", file.display(), e);
            }
        }
    }
    
    // Report violations
    if all_violations.is_empty() {
        log!("✓ All implementations are in correct order");
    } else {
        log!("✗ Implementation Order Violations:");
        log!("");
        log!("Standard trait impls (Eq, Debug, Display, etc.) should be AT THE BOTTOM (after custom trait impls).");
        log!("");
        
        for v in &all_violations {
            if let Ok(rel_path) = v.file.strip_prefix(&base_dir) {
                log!("{}:{}: Standard trait '{}' before custom trait '{}'", 
                         rel_path.display(), v.standard_line, v.standard_trait, v.custom_trait);
                log!("  Line {}: {} impl (standard trait)", v.standard_line, v.standard_trait);
                log!("  Line {}: {} impl (custom trait)", v.custom_line, v.custom_trait);
                log!("  → Standard trait impls should move to the bottom");
            }
        }
        
        log!("");
        log!("Correct order:");
        log!("  1. Data structure (struct/enum)");
        log!("  2. Trait definition");
        log!("  3. Inherent impl (impl Type {{ ... }})");
        log!("  4. Custom trait implementations");
        log!("  5. Standard trait implementations <- AT THE BOTTOM");
    }
    
    // Summary
    log!("");
    log!("Summary: {} files checked, {} violations", 
             format_number(files.len()), format_number(all_violations.len()));
    
    let elapsed = start.elapsed().as_millis();
    log!("Completed in {}ms", elapsed);
    
    if all_violations.is_empty() {
        Ok(())
    } else {
        std::process::exit(1);
    }
}

